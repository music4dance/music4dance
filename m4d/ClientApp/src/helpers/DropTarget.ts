import { ServiceMatcher } from "./ServiceMatcher";

const matcher: ServiceMatcher = new ServiceMatcher();

export async function checkServiceAndAdd(input: string, warn?: boolean): Promise<void> {
  const service = matcher.match(input);
  if (!service) {
    return;
  }
  const found = await checkService(input);

  if (!found) {
    let okay = false;
    if (warn) {
      // INT-TODO: Figure out if there is a way to do this with BSV-Next
      //   okay = await this.$bvModal.msgBoxConfirm(
      //     `It looks like you may have tried to search by ${service.name} id for a song not in the music4dance catalog.
      //  Would you like to add the song?`,
      //     {
      //       title: "Music Service Search?",
      //       okTitle: "Add Song",
      //     },
      //   );

      okay = window.confirm(
        `It looks like you may have tried to search by ${service.name} id for a song not in the music4dance catalog. Would you like to add the song?`,
      );
    }
    if (okay) {
      const id = matcher.parseId(input, service);
      window.location.href = `/song/augment?id=${id}`;
    }
  }
}

export async function checkServiceAndWarn(input: string): Promise<void> {
  return checkServiceAndAdd(input, true);
}

async function checkService(input: string): Promise<boolean> {
  const service = matcher.match(input);
  if (service) {
    try {
      const id = matcher.parseId(input, service);
      if (id) {
        const song = await matcher.findSong(input, true);
        if (song) {
          window.location.href = `/song/details/${song.songHistory.id}`;
          return true;
        }
      }
    } catch {
      // swallow any errors
    }
  }
  return false;
}
