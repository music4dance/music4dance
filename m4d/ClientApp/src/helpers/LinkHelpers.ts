import type { DanceInstance } from "@/models/DanceDatabase/DanceInstance";
import type { DanceType } from "@/models/DanceDatabase/DanceType";
import type { TempoRange } from "@/models/DanceDatabase/TempoRange";
import { wordsToKebab } from "./StringHelpers";

export function danceLink(dance: DanceInstance): string {
  return wordsToKebab(dance.shortName);
}

export function tempoLink(dance: DanceInstance | DanceType, tempo: TempoRange): string {
  return `/song/advancedsearch?dances=${dance.baseId}&tempomin=${tempo.min}&tempomax=${tempo.max}&sortorder=Dances`;
}

export function defaultTempoLink(dance: DanceInstance | DanceType): string {
  return tempoLink(dance, dance.tempoRange);
}

export function filteredTempoLink(dance: DanceInstance, filter: string): string {
  return tempoLink(dance, dance.filteredTempo([filter])!);
}
