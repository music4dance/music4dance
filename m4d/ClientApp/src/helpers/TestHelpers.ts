import { loadTestDances } from "./LoadTestDances";
// @ts-ignore
import tagDatabaseJson from "@/assets/tags.json";
declare global {
  interface Window {
    danceDatabaseJson?: string;
    tagDatabaseJson?: string;
  }
}

export const setupTestEnvironment = () => {
  window.danceDatabaseJson = loadTestDances();
  window.tagDatabaseJson = JSON.stringify(tagDatabaseJson);
};

export const mockResizObserver = () => {
  // @ts-ignore
  window.ResizeObserver = class ResizeObserver {
    observe() {
      // do nothing
    }
    unobserve() {
      // do nothing
    }
    disconnect() {
      // do nothing
    }
  };
};
