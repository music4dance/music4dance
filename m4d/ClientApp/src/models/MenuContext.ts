import type { AxiosInstance } from "axios";
import axios from "axios";

export interface MenuContextInterface {
  helpLink?: string;
  userName?: string;
  userId?: string;
  roles?: string[];
  indexId?: string;
  expiration?: Date;
  updateMessage?: string;
  marketingMessage?: string;
  xsrfToken?: string;
}

export class MenuContext implements MenuContextInterface {
  public helpLink?: string;
  public userName?: string;
  public userId?: string;
  public roles?: string[];
  public indexId?: string;
  public expiration?: Date;
  public updateMessage?: string;
  public marketingMessage?: string;
  public xsrfToken?: string;

  public constructor(init?: MenuContextInterface) {
    Object.assign(this, init);
  }

  public get isAdmin(): boolean {
    return !!this.roles?.find((r) => r === "dbAdmin");
  }

  public get isPremium(): boolean {
    return !!this.roles?.find((r) => r === "premium" || r === "trial");
  }

  public get canTag(): boolean {
    return !!this.roles?.find((r) => r === "canTag");
  }

  public get canEdit(): boolean {
    return !!this.roles?.find((r) => r === "canEdit");
  }

  public get isBeta(): boolean {
    return !!this.roles?.find((r) => r === "beta");
  }

  public get isAuthenticated(): boolean {
    return !!this.userName;
  }

  public get daysToExpiration(): number | undefined {
    if (!this.expiration) {
      return undefined;
    }
    const timeDiff = this.expiration.getTime() - Date.now();
    const dayDiff = timeDiff / (1000 * 60 * 60 * 24);
    return dayDiff > 0 ? dayDiff : 0;
  }

  public hasRole(role: string): boolean {
    return !!this.roles?.includes(role);
  }

  public getAccountLink(page: string): string {
    return `/identity/account/${page}?returnUrl=${window.location.pathname}?${window.location.search}`;
  }

  public get axiosXsrf(): AxiosInstance {
    return axios.create({
      headers: { RequestVerificationToken: this.xsrfToken },
    });
  }
}
