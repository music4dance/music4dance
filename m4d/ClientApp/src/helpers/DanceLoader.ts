import { DanceDatabase } from "@/models/DanceDatabase";

export function loadDancesFromString(s: string): DanceDatabase {
  const db = DanceDatabase.load(s);
  if (!db) {
    throw new Error("Dance Database not loaded as expected");
  }
  return db;
}
