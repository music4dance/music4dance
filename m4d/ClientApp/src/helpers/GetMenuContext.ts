import { MenuContext } from "@/models/MenuContext";
import type { AxiosInstance } from "axios";

declare global {
  interface Window {
    menuContext: MenuContext;
  }
}

let menuContextObject: MenuContext | null = null;

export function getMenuContext(): MenuContext {
  if (!menuContextObject) {
    menuContextObject = window.menuContext
      ? new MenuContext(window.menuContext)
      : new MenuContext();
  }
  return menuContextObject;
}

export function getAxiosXsrf(): AxiosInstance {
  return getMenuContext().axiosXsrf;
}
