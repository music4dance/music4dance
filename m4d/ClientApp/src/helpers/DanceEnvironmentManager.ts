import { DanceDatabase } from "@/models/DanceDatabase/DanceDatabase";
import { loadDancesFromString } from "./DanceLoader";

declare global {
  interface Window {
    danceDatabaseJson?: string;
    danceDatabase: DanceDatabase | undefined;
  }
}

export function safeDanceDatabase(): DanceDatabase {
  if (window.danceDatabase) {
    return window.danceDatabase;
  }

  if (!window.danceDatabaseJson) {
    throw new Error("danceDatabaseJson not defined on window");
  }

  const db = loadDancesFromString(window.danceDatabaseJson);
  if (db) {
    window.danceDatabase = db;
    return db;
  }
  throw new Error("Dance Database not loaded as expected");
}
