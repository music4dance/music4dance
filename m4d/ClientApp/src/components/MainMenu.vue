<script setup lang="ts">
import { MenuContext } from "@/models/MenuContext";
import { computed, ref } from "vue";
import { checkServiceAndWarn } from "@/helpers/DropTarget";
import logo from "@/assets/images/header-logo.png";
import dancers from "@/assets/images/swing-ui.png";

const renewalTag = "renewal-acknowledged";
const marketingTag = "marketing-acknowledged";

const props = defineProps<{ context: MenuContext }>();
const searchString = ref<string>("");

const renewal = ref(!sessionStorage.getItem(renewalTag));
const marketing = ref(!sessionStorage.getItem(marketingTag));

const songIndex = computed(() => {
  const context = props.context;
  return context.isAdmin && context.indexId ? ` (${context.indexId})` : "";
});

const helpLink = computed(() => {
  return props.context.helpLink ?? "https://music4dance.blog/music4dance-help/";
});

const songLink = computed(() => {
  return `/users/info/${props.context.userName}`;
});

const searchLink = computed(() => {
  return `/searches/?user=${props.context.userName}`;
});

const showExpiration = computed(() => {
  return (
    props.context.daysToExpiration !== undefined &&
    props.context.daysToExpiration < 30 &&
    !sessionStorage.getItem(renewalTag)
  );
});

const showMarketing = computed(() => {
  return (
    props.context.marketingMessage &&
    props.context.marketingMessage.length > 0 &&
    !sessionStorage.getItem(marketingTag)
  );
});

const isTest = computed(() => {
  return window.location.hostname.endsWith(".azurewebsites.net");
});

function accountLink(type: string): string {
  const url = window.location.pathname + window.location.search;
  return `/identity/account/${type}?returnUrl=${url}`;
}

function onDismissed(target: string): void {
  sessionStorage.setItem(target, "true");
}

function search(): void {
  event?.preventDefault();
  window.location.href = `/search?search=${encodeURIComponent(searchString.value)}`;
}
</script>

<template>
  <div>
    <BAlert v-if="isTest" :model-value="true" variant="warning" style="margin-bottom: 0"
      >This is a TEST site. Please navigate to
      <a href="https://www.music4dance.net" class="alert-link">wwww.music4dance.net</a>
      to use the production site.</BAlert
    >
    <BAlert
      v-if="context.updateMessage"
      v-model="renewal"
      variant="danger"
      style="margin-bottom: 0"
      dismissable
      >{{ context.updateMessage }}</BAlert
    >
    <BNavbar id="mainMenu" data-bs-theme="dark" variant="primary" toggleable="lg">
      <BNavbarBrand href="/">
        <img :src="logo" height="40" title="music4dance" />
      </BNavbarBrand>

      <BNavbarToggle id="drop-toggle" target="nav-collapse"></BNavbarToggle>

      <BCollapse id="nav-collapse" is-nav>
        <BNavbarNav>
          <BNavItemDropdown id="music-menu" text="Music">
            <BDropdownItem href="/dances">Dances</BDropdownItem>
            <BDropdownItem href="/dances/ballroom-competition-categories" class="nav-subitem"
              >&nbsp;&nbsp;Ballroom</BDropdownItem
            >
            <BDropdownItem href="/dances/latin" class="nav-subitem"
              >&nbsp;&nbsp;Latin</BDropdownItem
            >
            <BDropdownItem href="/dances/swing" class="nav-subitem"
              >&nbsp;&nbsp;Swing</BDropdownItem
            >
            <BDropdownItem href="/dances/tango" class="nav-subitem"
              >&nbsp;&nbsp;Tango</BDropdownItem
            >
            <BDropdownItem href="/song">Song Library</BDropdownItem>
            <BDropdownItem href="/song/advancedsearchform" class="nav-subitem"
              >&nbsp;&nbsp;Advanced Search</BDropdownItem
            >
            <BDropdownItem href="/song/augment" class="nav-subitem"
              >&nbsp;&nbsp;Add Song</BDropdownItem
            >
            <BDropdownItem href="/song/newmusic" class="nav-subitem"
              >&nbsp;&nbsp;New Music</BDropdownItem
            >
            <BDropdownItem href="/dances/wedding-music">Wedding</BDropdownItem>
            <BDropdownItem href="/song/holidaymusic">Holiday</BDropdownItem>
            <BDropdownItem href="/song/holidaymusic?occassion=halloween">Halloween</BDropdownItem>
            <BDropdownItem href="/tag">Tags</BDropdownItem>
          </BNavItemDropdown>
          <BNavItemDropdown id="tools-menu" text="Tools">
            <BDropdownItem href="/home/counter">Tempo Counter</BDropdownItem>
            <BDropdownItem href="/home/tempi">Tempi (Tempos)</BDropdownItem>
            <BDropdownItem href="/song/advancedsearchform">Advanced Search</BDropdownItem>
            <BDropdownItem v-if="context.isBeta || context.isAdmin" href="/home/spotifyexplorer"
              >Spotify Explorer (BETA)</BDropdownItem
            >
          </BNavItemDropdown>
          <BNavItemDropdown id="info" text="Info">
            <BDropdownItem :href="helpLink">Help</BDropdownItem>
            <BDropdownItem href="https://music4dance.blog/">Blog</BDropdownItem>
            <BDropdownItem href="/home/faq">FAQ</BDropdownItem>
            <BDropdownItem href="/home/about">About</BDropdownItem>
            <BDropdownItem href="/home/readinglist">Reading List</BDropdownItem>
            <BDropdownItem href="/home/sitemap">Site Map</BDropdownItem>
            <BDropdownItem href="/home/privacypolicy">Privacy Policy</BDropdownItem>
            <BDropdownItem href="/home/credits">Credits</BDropdownItem>
          </BNavItemDropdown>
          <BNavItem id="contribute-menu" right href="/home/contribute">Contribute</BNavItem>
          <BNavItemDropdown id="admin-menu" text="Admin" v-if="context.isAdmin">
            <BDropdownItem href="/admin">Index</BDropdownItem>
            <BDropdownItem href="/applicationusers">Users</BDropdownItem>
            <BDropdownItem href="/activitylog">Activity Log</BDropdownItem>
            <BDropdownItem href="/tag/list">Tags</BDropdownItem>
            <BDropdownItem href="/admin/diagnostics">Diagnostics</BDropdownItem>
            <BDropdownItem href="/admin/initializationTasks"
              >Initialization and Cleanup</BDropdownItem
            >
            <BDropdownItem href="/admin/scraping">Scraping</BDropdownItem>
            <BDropdownItem href="/playlist">PlayLists</BDropdownItem>
            <BDropdownItem href="/Searches?showDetails=True&user=all">Searches</BDropdownItem>
            <BDropdownItem href="/song/rawsearchform">Raw Search</BDropdownItem>
            <BDropdownItem href="/admin/uploadbackup">Uploads and Backups</BDropdownItem>
          </BNavItemDropdown>
        </BNavbarNav>

        <BNavbarNav class="ms-auto">
          <BNavForm id="search" @submit="search">
            <BFormInput
              id="search-text"
              v-model="searchString"
              size="sm"
              class="me-sm-2"
              placeholder="Search"
              @input="checkServiceAndWarn"
            ></BFormInput>
            <BButton
              size="sm"
              class="mx-2 my-2 my-sm-0"
              type="submit"
              :disabled="!searchString.trim()"
              >Search</BButton
            >
          </BNavForm>
          <template v-if="context.userName">
            <BNavItemDropdown>
              <template #button-content>
                {{ context.userName }} {{ songIndex }}
                <img :src="dancers" alt="User Icon" height="30" width="30" />
              </template>
              <BDropdownItem right href="/identity/account/manage">My Profile</BDropdownItem>
              <BDropdownItem right :href="songLink">My Songs</BDropdownItem>
              <BDropdownItem right :href="searchLink">My Searches</BDropdownItem>
              <BDropdownItem right href="javascript:document.getElementById('logoutForm').submit()"
                >Log out</BDropdownItem
              >
            </BNavItemDropdown>
          </template>
          <template v-else>
            <BNavItem right :href="accountLink('register')">Register</BNavItem>
            <BNavItem right :href="accountLink('login')">Login</BNavItem>
          </template>
        </BNavbarNav>
      </BCollapse>
    </BNavbar>
    <BAlert
      v-if="showMarketing"
      v-model="marketing"
      variant="success"
      dismissible
      style="margin-bottom: 0"
      @update:modelValue="onDismissed(marketingTag)"
      ><span v-html="context.marketingMessage"></span
    ></BAlert>
    <BAlert
      v-if="showExpiration"
      variant="warning"
      v-model="renewal"
      dismissible
      style="margin-bottom: 0"
      @update:modelValue="onDismissed(renewalTag)"
    >
      Your premium subcription will expire in
      {{ Math.round(context.daysToExpiration || 0) }}
      day(s). Please
      <a href="/home/contribute" class="alert-link">click here</a> to renew. Thanks for your
      continued support.
    </BAlert>
    <form id="logoutForm" action="/identity/account/logout" method="post" style="height: 0">
      <input name="__RequestVerificationToken" type="hidden" :value="context.xsrfToken" />
      <input type="hidden" name="returnUrl" value="/" />
      <button id="logout" type="submit" class="btn btn-link"></button>
    </form>
  </div>
</template>
