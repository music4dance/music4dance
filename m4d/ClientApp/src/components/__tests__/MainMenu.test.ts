import { mount } from "@vue/test-utils";
import { describe, expect, test } from "vitest";
import { MenuContext } from "@/models/MenuContext";
import MainMenu from "../MainMenu.vue";
import { mockResizObserver } from "@/helpers/TestHelpers";
import { modalControllerPlugin, modalManagerPlugin } from "bootstrap-vue-next";

describe("MainMenu.vue", () => {
  test("Renders MainMenu for an anonymous user", () => {
    mockResizObserver();
    const context = new MenuContext();
    const wrapper = mount(MainMenu, {
      props: { context },
      global: { plugins: [modalControllerPlugin, modalManagerPlugin] },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
