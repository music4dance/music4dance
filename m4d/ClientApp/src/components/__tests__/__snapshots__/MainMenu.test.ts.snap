// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainMenu.vue > Renders MainMenu for an anonymous user 1`] = `
"<div>
  <!--teleport start-->
  <!--teleport end-->
  <!--teleport start-->
  <!--teleport end-->
  <!--v-if-->
  <!--v-if-->
  <nav class=\\"navbar bg-primary navbar-expand-lg\\" id=\\"mainMenu\\" data-bs-theme=\\"dark\\">
    <div class=\\"container-fluid\\"><a class=\\"navbar-brand\\" replace=\\"false\\" href=\\"/\\"><img src=\\"/vclient/src/assets/images/header-logo.png\\" height=\\"40\\" title=\\"music4dance\\"></a><button class=\\"navbar-toggler collapsed\\" type=\\"button\\" aria-label=\\"Toggle navigation\\" id=\\"drop-toggle\\" aria-controls=\\"nav-collapse\\" aria-expanded=\\"false\\"><span class=\\"navbar-toggler-icon\\"></span></button>
      <div id=\\"nav-collapse\\" class=\\"collapse navbar-collapse\\" is-nav=\\"true\\">
        <ul class=\\"navbar-nav\\">
          <li class=\\"nav-item dropdown\\">
            <div class=\\"dropdown\\"><button class=\\"btn btn-md btn-link nav-link dropdown-toggle\\" type=\\"button\\" id=\\"music-menu\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\">Music</button>
              <!---->
              <!--teleport start-->
              <ul style=\\"position: absolute; left: 0px; top: 0px; display: none;\\" class=\\"dropdown-menu overflow-auto\\" aria-labelledby=\\"music-menu\\" role=\\"menu\\">
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/dances\\" role=\\"menuitem\\">Dances</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/dances/ballroom-competition-categories\\" role=\\"menuitem\\">&nbsp;&nbsp;Ballroom</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/dances/latin\\" role=\\"menuitem\\">&nbsp;&nbsp;Latin</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/dances/swing\\" role=\\"menuitem\\">&nbsp;&nbsp;Swing</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/dances/tango\\" role=\\"menuitem\\">&nbsp;&nbsp;Tango</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/dances/country\\" role=\\"menuitem\\">&nbsp;&nbsp;Country</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/song\\" role=\\"menuitem\\">Song Library</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/song/advancedsearchform\\" role=\\"menuitem\\">&nbsp;&nbsp;Advanced Search</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/song/augment\\" role=\\"menuitem\\">&nbsp;&nbsp;Add Song</a></li>
                <li role=\\"presentation\\" class=\\"nav-subitem\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/song/newmusic\\" role=\\"menuitem\\">&nbsp;&nbsp;New Music</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/dances/wedding-music\\" role=\\"menuitem\\">Wedding</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/song/holidaymusic\\" role=\\"menuitem\\">Holiday</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/song/holidaymusic?occassion=halloween\\" role=\\"menuitem\\">Halloween</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/tag\\" role=\\"menuitem\\">Tags</a></li>
              </ul>
              <!--teleport end-->
            </div>
          </li>
          <li class=\\"nav-item dropdown\\">
            <div class=\\"dropdown\\"><button class=\\"btn btn-md btn-link nav-link dropdown-toggle\\" type=\\"button\\" id=\\"tools-menu\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\">Tools</button>
              <!---->
              <!--teleport start-->
              <ul style=\\"position: absolute; left: 0px; top: 0px; display: none;\\" class=\\"dropdown-menu overflow-auto\\" aria-labelledby=\\"tools-menu\\" role=\\"menu\\">
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/counter\\" role=\\"menuitem\\">Tempo Counter</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/tempi\\" role=\\"menuitem\\">Tempi (Tempos)</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/song/advancedsearchform\\" role=\\"menuitem\\">Advanced Search</a></li>
                <!--v-if-->
              </ul>
              <!--teleport end-->
            </div>
          </li>
          <li class=\\"nav-item dropdown\\">
            <div class=\\"dropdown\\"><button class=\\"btn btn-md btn-link nav-link dropdown-toggle\\" type=\\"button\\" id=\\"info\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\">Info</button>
              <!---->
              <!--teleport start-->
              <ul style=\\"position: absolute; left: 0px; top: 0px; display: none;\\" class=\\"dropdown-menu overflow-auto\\" aria-labelledby=\\"info\\" role=\\"menu\\">
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"https://music4dance.blog/music4dance-help/\\" role=\\"menuitem\\">Help</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"https://music4dance.blog/\\" role=\\"menuitem\\">Blog</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/faq\\" role=\\"menuitem\\">FAQ</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/about\\" role=\\"menuitem\\">About</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/readinglist\\" role=\\"menuitem\\">Reading List</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/sitemap\\" role=\\"menuitem\\">Site Map</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/privacypolicy\\" role=\\"menuitem\\">Privacy Policy</a></li>
                <li role=\\"presentation\\"><a class=\\"dropdown-item\\" replace=\\"false\\" href=\\"/home/credits\\" role=\\"menuitem\\">Credits</a></li>
              </ul>
              <!--teleport end-->
            </div>
          </li>
          <li class=\\"nav-item\\" id=\\"contribute-menu\\" right=\\"\\"><a class=\\"nav-link\\" replace=\\"false\\" href=\\"/home/contribute\\">Contribute</a></li>
          <!--v-if-->
        </ul>
        <ul class=\\"navbar-nav ms-auto\\">
          <li class=\\"d-flex flex-row align-items-center flex-wrap\\">
            <form id=\\"search\\" class=\\"d-flex\\">
              <div class=\\"input-group\\" role=\\"group\\">
                <!---->
                <!--v-if--><input id=\\"search-text\\" class=\\"form-control form-control-sm\\" type=\\"text\\" placeholder=\\"Search...\\" autocomplete=\\"off\\" list=\\"search-text-suggestions\\" value=\\"\\"><datalist id=\\"search-text-suggestions\\"></datalist><button class=\\"btn link-undefined link-underline-undefined btn-sm btn-outline-secondary\\" type=\\"button\\"><svg viewBox=\\"0 0 16 16\\" width=\\"1.2em\\" height=\\"1.2em\\">
                    <path fill=\\"currentColor\\" d=\\"M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001q.044.06.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1 1 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0a5.5 5.5 0 0 1 11 0\\"></path>
                  </svg></button>
                <!---->
              </div>
            </form>
          </li>
          <li class=\\"nav-item\\" right=\\"\\"><a class=\\"nav-link\\" replace=\\"false\\" href=\\"/identity/account/register?returnUrl=/\\">Register</a></li>
          <li class=\\"nav-item\\" right=\\"\\"><a class=\\"nav-link\\" replace=\\"false\\" href=\\"/identity/account/login?returnUrl=/\\">Login</a></li>
        </ul>
      </div>
    </div>
  </nav>
  <!--v-if-->
  <!--v-if-->
  <form id=\\"logoutForm\\" action=\\"/identity/account/logout\\" method=\\"post\\" style=\\"height: 0px;\\"><input name=\\"__RequestVerificationToken\\" type=\\"hidden\\"><input type=\\"hidden\\" name=\\"returnUrl\\" value=\\"/\\"><button id=\\"logout\\" type=\\"submit\\" class=\\"btn btn-link\\"></button></form>
</div>"
`;
